third_party/move/move-model/src/builder/model_builder.rs

```diff
@@ -387,6 +390,76 @@ impl<'env> ModelBuilder<'env> {
             .expect("invalid Type::Struct")
     }
 
+    // assuming all structs are well defined
+    // checks whether the type is instantiated properly, i.e.,
+    // - the type arguments given to the struct are instantiated properly
+    // - the type arguments satisfy the ability constraints defined on the struct generics
+    // and returns the abilities of the given type
+    // `ty_params` specify the abilities held by type params
+    pub fn check_instantiation(
+        &self,
+        ty: &Type,
+        ty_params: &[TypeParameter],
+        loc: &Loc,
+    ) -> AbilitySet {
+        check_instantiation(
```

Comment:
Please don't use the name `check_instantiation` for multiple things.  This looks like a recursive call at first.  Please be more specific with names.  Actually, since this function always does the check, renaming the one called here (which optionally does the check) makes sense.

Reply:
done